File: server/apps/assetmanager/schemas/entity_schemas/syndicate_schemas.py
"""
Syndicate Schemas

Pydantic schemas for the Syndicate model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime

# ==========================================
# Syndicate Schema (Full Representation)
# ==========================================

class Syndicate(BaseModel):
    """Syndicate schema - full representation"""
    id: int
    entity_id: int = Field(description="Managing entity ID")
    name: str = Field(min_length=1, max_length=255, description="Syndicate name")
    carried_interest_percentage: float | None = Field(None, description="Carried interest percentage")
    minimum_investment: float | None = Field(None, description="Minimum investment amount")
    maximum_investment: float | None = Field(None, description="Maximum investment amount")
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class CreateSyndicate(BaseModel):
    """Schema for creating a new syndicate"""
    entity_id: int = Field(description="Managing entity ID")
    name: str = Field(min_length=1, max_length=255, description="Syndicate name")
    carried_interest_percentage: float | None = Field(None, description="Carried interest percentage")
    minimum_investment: float | None = Field(None, description="Minimum investment amount")
    maximum_investment: float | None = Field(None, description="Maximum investment amount")

class UpdateSyndicate(BaseModel):
    """Schema for updating a syndicate"""
    entity_id: int | None = None
    name: str | None = Field(None, min_length=1, max_length=255)
    carried_interest_percentage: float | None = None
    minimum_investment: float | None = None
    maximum_investment: float | None = None

# ==========================================
# Response Types
# ==========================================

class SyndicateResponse(BaseModel):
    """Response containing a single syndicate"""
    success: bool
    data: Syndicate | None = None
    error: str | None = None

class SyndicatesResponse(BaseModel):
    """Response containing multiple syndicates"""
    success: bool
    data: list[Syndicate] | None = None
    error: str | None = None

File: client/src/modules/assetmanager/schemas/entity/syndicate.schemas.ts
/**
 * Syndicate Schemas
 *
 * Zod validation schemas for Syndicate model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/entity_models.py
 * - Schema: /server/apps/assetmanager/schemas/entity_schemas/syndicate_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/entity_subrouters/syndicate_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// Syndicate Schema (Full Representation)
// ==========================================

/**
 * Syndicate schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class Syndicate(BaseModel)
 */
export const SyndicateSchema = z.object({
  id: z.number(),
  entity_id: z.number(),
  name: z.string().min(1, 'Name is required').max(255, 'Name must be less than 255 characters'),
  carried_interest_percentage: z.number().nullable(),
  minimum_investment: z.number().nullable(),
  maximum_investment: z.number().nullable(),
  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new syndicate (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class CreateSyndicate(BaseModel)
 */
export const CreateSyndicateSchema = z.object({
  entity_id: z.number(),
  name: z.string().min(1, 'Name is required').max(255, 'Name must be less than 255 characters'),
  carried_interest_percentage: z.number().nullable().optional(),
  minimum_investment: z.number().nullable().optional(),
  maximum_investment: z.number().nullable().optional(),
});

/**
 * Schema for updating a syndicate (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class UpdateSyndicate(BaseModel)
 */
export const UpdateSyndicateSchema = z.object({
  entity_id: z.number().nullable().optional(),
  name: z.string().min(1, 'Name is required').max(255, 'Name must be less than 255 characters').optional(),
  carried_interest_percentage: z.number().nullable().optional(),
  minimum_investment: z.number().nullable().optional(),
  maximum_investment: z.number().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type Syndicate = z.infer<typeof SyndicateSchema>;
export type CreateSyndicate = z.infer<typeof CreateSyndicateSchema>;
export type UpdateSyndicate = z.infer<typeof UpdateSyndicateSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single syndicate
 * Backend equivalent: class SyndicateResponse(BaseModel)
 */
export type SyndicateResponse = {
  success: boolean;
  data?: Syndicate;
  error?: string;
};

/**
 * Response containing multiple syndicates
 * Backend equivalent: class SyndicatesResponse(BaseModel)
 */
export type SyndicatesResponse = {
  success: boolean;
  data?: Syndicate[];
  error?: string;
};
